### PR TITLE
Feature/jwt

### PR DESCRIPTION
**Pull request by Ivan-Marquez**
# JWT
> Adding JWT refresh token flow.

## Change log:
+ Updated express headers configuration.
+ Added validation to check JWT remaining time.
+ Added custom header with updated token.
+ Updated tests (100% code coverage).

### Additional changes:
+ Updated logger configuration. Removed max files property.

### Pending changes:
+ Create script to delete log files after a certain period of time.

### CODE REVIEW

File: src/config/express.js [hunk 1]
```diff
@@ -47,11 +47,14 @@ module.exports = () => {
...
   app.use((req, res, next) => {
+    const allowedHeaders = ['Origin', 'X-Requested-With', 'Content-Type', 'Accept', 'Authorization',
+      'Access-Control-Allow-Credentials']
+
     res.header('Access-Control-Allow-Origin', '*')
-    res.header('Access-Control-Allow-Methods', 'PUT, GET, POST, DELETE, OPTIONS')
-    res.header('Access-Control-Allow-Headers',
-      'Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Credentials')
     res.header('Access-Control-Allow-Credentials', 'true')
+    res.header('Access-Control-Allow-Methods', 'PUT, GET, POST, DELETE, OPTIONS')
```
Comment (kevinpr007) [inline src/config/express.js]: Do you can do the same approach with also this variables?

File: src/config/logger.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ const serializer = data => {
 
   return `db.${data.coll}.${data.method}(${query}, ${options})`
 }
-
+// TODO: create script to delete files after a certain period of time.
```
Comment (kevinpr007) [inline src/config/logger.js]: I think this is not a programming process. This is a management/administrator process.

File: src/config/strategies/jwt.js [hunk 3]
```diff
@@ -2,24 +2,39 @@ const config = require('../config')
...
-      if (!user) {
-        return cb(null, false, {
-          message: 'Unknown user'
-        })
-      } else {
-        return cb(null, user)
-      }
-    }).catch(err => cb(err))
+  passport.use(new JwtStrategy(jwtOptions, (req, payload, cb) => {
+    const expDate = new Date(parseInt(`${payload.exp}000`))
```
Comment (kevinpr007) [inline src/config/strategies/jwt.js]: Why we have 000 hardcode?

File: src/config/strategies/jwt.js [hunk 4]
```diff
@@ -2,24 +2,39 @@ const config = require('../config')
...
+        if (getRemainingTime(expDate) <= parseInt(config.EXP_SECONDS)) {
+          return refreshToken(user)
+          .then(jwt => {
+            req.res.set('X-Updated-JWT', jwt)
+            return Promise.resolve(cb(null, user))
+          })
+        } else {
+          return Promise.resolve(cb(null, user))
+        }
+      }).catch(err => cb(err))
```
Comment (kevinpr007) [inline src/config/strategies/jwt.js]: Can we remove the lines "return Promise.resolve(cb(null, user))" and the else and then write return Promise.resolve(cb(null, user)) outside of the if?

File: src/controllers/auth.controller.js [hunk 5]
```diff
@@ -44,7 +47,10 @@ module.exports = (authService, userService, emailService) => {
         return userService.registerUser(req.body)
       })
       .then(user => {
-        user._doc.token = authService.getToken(user)
+        let token = authService.getToken(user)
```
Comment (kevinpr007) [inline src/controllers/auth.controller.js]: Why we are using the this lines twice?

File: src/services/auth.service.js [hunk 6]
```diff
@@ -5,27 +5,40 @@ const mongoose = require('mongoose')
...
   const generateToken = user =>
     jwt.sign(user, config.SECRET, {
-      expiresIn: parseInt(tokenLife)
+      expiresIn: parseInt(tokenExpTime)
     })
 
   const getToken = user => `JWT ${generateToken(user)}`
 
+  const setExpirationDate = () =>
+    moment().add((parseInt(tokenExpTime) / 3600), 'h')
```
Comment (kevinpr007) [inline src/services/auth.service.js]: What does mean 3600?

File: src/services/auth.service.js [hunk 7]
```diff
@@ -5,27 +5,40 @@ const mongoose = require('mongoose')
...
     const date = new Date()
     user = _.extend(user, User)
-    return randomBytes(parseInt(config.RANDOM_BYTES)).then(buffer => {
-      user.resetPasswordToken = buffer.toString('hex')
-      user.resetPasswordExpires = date.setSeconds(date.getSeconds() + parseInt(tokenLife))
-      return user.save()
-    })
+    return randomBytes(parseInt(config.RANDOM_BYTES))
+      .then(buffer => {
+        user.resetPasswordToken = buffer.toString('hex')
```
Comment (kevinpr007) [inline src/services/auth.service.js]: What does mean this line?

File: src/util/util.helpers.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @desc Returns remaining time (in seconds) of date passed as argument.
+ * @param {Date} endtime
+ * @returns {number}
+ */
+const getRemainingTime = endTime => {
+  var startDate = new Date()
+  return (endTime.getTime() - startDate.getTime()) / 1000
```
Comment (kevinpr007) [inline src/util/util.helpers.js]: What does mean 1000?

File: src/services/auth.service.js (hunk 7 above)
Comment (Ivan-Marquez) [inline src/services/auth.service.js]: this line was already in place before this feature. It's just for encryption purposes.

File: src/util/util.helpers.js (hunk 8 above)
Comment (Ivan-Marquez) [inline src/util/util.helpers.js]: that's how you calculate the remaining time (in seconds).

File: src/services/auth.service.js (hunk 6 above)
Comment (Ivan-Marquez) [inline src/services/auth.service.js]: We're setting `tokenExpTime` in seconds. That's the formula to convert seconds to hours. This function adds `tokenExpTime` (in hours) to the current date.

File: src/config/logger.js (hunk 2 above)
Comment (Ivan-Marquez) [inline src/config/logger.js]: Whether it's automated or not. It's still pending.

Comment (kevinpr007) [inline src/config/logger.js]: Ok

File: src/config/strategies/jwt.js (hunk 3 above)
Comment (Ivan-Marquez) [inline src/config/strategies/jwt.js]: We'll talk about this.

File: src/services/auth.service.js (hunk 7 above)
Comment (Ivan-Marquez) [inline src/services/auth.service.js]: This method is used to encode each byte of the buffer as two hexadecimal characters.

File: src/services/auth.service.js [hunk 9]
```diff
@@ -1,44 +1,58 @@
...
 mongoose.Promise = Promise
 
 const User = require('../models/user.model')
-const tokenLife = config.TOKEN_LIFE
+const tokenExpTime = parseInt(TOKEN_EXP_TIME)
+const moment = require('moment')
 const _ = require('lodash')
 
 const randomBytes = Promise.promisify(require('crypto').randomBytes)
+const HOURS = 3600
```
Comment (kevinpr007) [inline src/services/auth.service.js]: Perfect!

File: src/services/auth.service.js [hunk 10]
```diff
@@ -1,44 +1,58 @@
...
   const generateToken = user =>
-    jwt.sign(user, config.SECRET, {
-      expiresIn: parseInt(tokenLife)
+    jwt.sign(user, SECRET, {
+      expiresIn: tokenExpTime
     })
 
   const getToken = user => `JWT ${generateToken(user)}`
 
+  const setExpirationDate = () => moment().add((tokenExpTime / HOURS), 'h')
```
Comment (kevinpr007) [inline src/services/auth.service.js]: Excellent!

File: src/util/util.helpers.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+/**
```
Comment (kevinpr007) [inline src/util/util.helpers.js]: This file looks very good!!!!

**Pull request merged**